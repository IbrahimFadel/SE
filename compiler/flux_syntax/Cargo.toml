[package]
name = "flux_syntax"
version = "0.1.0"
edition = "2021"

[dependencies]
cstree = "0.12.0"
text-size = "1.1.1"

flux_lexer = { path = "../flux_lexer" }
salsa = { path = "../../lib/salsa" }
