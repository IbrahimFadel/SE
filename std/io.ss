fn print() -> i32 {
    return 0;
}