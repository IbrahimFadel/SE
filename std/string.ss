struct string {
    i8 *buffer;
    i64 length;
    i64 maxLength;
    i64 factor;
};

fn string_create_default(string *this) -> void {
    this->buffer = nullptr;
    this->length = 0;
    this->maxLength = 0;
    this->factor = 16;
}

fn string_delete(string *this) -> void {
    i8 *buf = this->buffer;
    if(buf != nullptr) {
        free(buf);
    }
}

fn string_resize(string *this, i64 value) -> void {
    i8 *output = malloc(value);

    i8 *buf = this->buffer;
    i64 len = this->length;

    memcpy(output, buf, len);

    free(buf);
    this->buffer = output;
}

fn string_add_char(string *this, i8 value) -> void {
    i64 len = this->length;
    i64 maxLen = this->maxLength;

    if(len == maxLen) {
        i64 factor = this->factor;
        string_resize(this, maxLen + factor);
    }

    i8 *buf = this->buffer;
    buf[len] = value;
    i64 newLen = len + 1;
    this->length = newLen;
}