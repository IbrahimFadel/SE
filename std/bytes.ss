import "./io.ss";

fn bytesFunction() -> void {
    
}