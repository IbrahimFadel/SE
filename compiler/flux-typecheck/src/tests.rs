use codespan_reporting::term::termcolor::WriteColor;
use flux_syntax::{ast, ast::AstNode};
use smol_str::SmolStr;
use std::{collections::HashMap, str};

struct ErrorStream {
	s: String,
}

impl WriteColor for ErrorStream {
	fn supports_color(&self) -> bool {
		false
	}

	fn is_synchronous(&self) -> bool {
		false
	}

	fn reset(&mut self) -> std::io::Result<()> {
		Ok(())
	}

	fn set_color(
		&mut self,
		_: &codespan_reporting::term::termcolor::ColorSpec,
	) -> std::io::Result<()> {
		Ok(())
	}
}

impl std::io::Write for ErrorStream {
	fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
		let s = match str::from_utf8(buf) {
			Ok(v) => v,
			Err(e) => panic!("Invalid UTF-8 sequence: {}", e),
		};
		self.s += s;
		Ok(buf.len())
	}

	fn flush(&mut self) -> std::io::Result<()> {
		Ok(())
	}
}

#[macro_export]
#[cfg(test)]
macro_rules! test_typeinf {
	($name:ident, $src:literal) => {
		paste::paste! {
				#[test]
				fn [<test_parse_ $name>]() {
					let mut err_reporting = flux_error::FluxErrorReporting::default();
					let file_id = err_reporting
						.add_file(format!("main"), $src.to_string())
						.expect("could not add file");
					let cst = flux_parser::parse($src, file_id);
					assert!(cst.errors.is_empty());
					let root = ast::Root::cast(cst.syntax()).unwrap();
					let (mut hir_module, _) = flux_hir::lower(vec![SmolStr::from("foo")], root, file_id);
					let function_exports = HashMap::new();
					let type_exports = HashMap::new();
					let res = crate::typecheck_hir_module(&mut hir_module, &function_exports, &type_exports);
					let mut s = String::new();
					if let Some(err) = res.err() {
						let mut writer = ErrorStream {
							s: String::new()
						};
						let _ = codespan_reporting::term::emit(&mut writer, &codespan_reporting::term::Config::default(), &err_reporting.files, &err.to_diagnostic());
						s = writer.s;
					} else {
						for f in &hir_module.functions {
							if let Some(name) = &f.name {
								s += &format!("{}\n--------------------\n", name.node);
								for stmt in &f.block.0 {
									if let Some(stmt) = stmt {
										if let flux_hir::Stmt::VarDecl(var) = &stmt.node {
											s += &format!("{} -> {:?}\n", var.name, var.ty.node);
										}
									}
								}
							}
						}
					}
					let mut settings = insta::Settings::clone_current();
					settings.set_snapshot_path("./snapshots");
					settings.bind(|| {
						insta::assert_snapshot!(s);
					});
				}
		}
	};
}

test_typeinf!(
	int_defaults_u32,
	r#"fn main() {
  let x = 0;
}"#
);

test_typeinf!(
	basic_dependency,
	r#"fn main() {
  i16 x = 0;
	let y = x;
}"#
);

test_typeinf!(
	basic_backwards_dependency,
	r#"fn main() {
  let x = 0;
	i16 y = x;
}"#
);

test_typeinf!(
	function_call_type_propagation,
	r#"fn foo(i13 x) {

}

fn main() {
	let x = 0;
	let y = x;
	foo(y);
}"#
);

test_typeinf!(
	function_call_incorrect_num_args,
	r#"fn foo(i13 x) {

}

fn main() {
	foo();
}"#
);

test_typeinf!(
	function_call_incorrect_arg_types,
	r#"fn foo(i13 x) {

}

fn main() {
	i32 y = 0;
	foo(y);
}"#
);

test_typeinf!(
	function_call_type_propagation_to_addition,
	r#"fn foo(i13 x) {

}

fn main() {
	let x = 0;
	let y = x;
	i32 a = 1;
	foo(y);
	let z = y + a;
}"#
);

test_typeinf!(
	function_call_type_propagation_throughout_scopes,
	r#"fn foo(i13 x) {
	
}

fn main() {
	let x = 0;
	let y = x;
	foo(y);

	if x == 0 {
		let z = y;
		i32 y = 10;
		if z == y {

		}
	}
}"#
);
