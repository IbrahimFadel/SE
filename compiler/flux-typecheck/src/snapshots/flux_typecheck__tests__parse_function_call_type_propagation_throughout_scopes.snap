---
source: compiler/flux-typecheck/src/tests.rs
expression: s
---
error[E0000]: could not unify `i13` and `i32`
   ┌─ main:13:6
   │
 6 │     let x = 0;
   │             - `i13` type
   ·
12 │         i32 y = 10;
   │         ---- `i32` type
13 │         if z == y {
   │            ^^^^^^^ could not unify `i13` and `i32`


